src/vsr/replica.zig

```diff
@@ -2394,10 +2406,79 @@ pub fn Replica(
+            assert(self.committing);
+            assert(self.commit_callback != null);
+            assert(self.op_checkpoint == self.superblock.staging.vsr_state.commit_min);
+            assert(self.op_checkpoint == self.superblock.working.vsr_state.commit_min);
+
+            const op = self.commit_prepare.?.header.op;
+            if (op == self.op_checkpoint_trigger()) {
+                assert((op + 1) % config.lsm_batch_multiple == 0);
+                self.state_machine.forest.checkpoint(commit_op_checkpoint_forest_callback, op);
+            } else {
+                assert(op < self.op_checkpoint_trigger());
+                self.commit_op_done();
+            }
+        }
+
+        fn commit_op_checkpoint_forest_callback(forest: *StateMachine.Forest) void {
+            const state_machine = @fieldParentPtr(StateMachine, "forest", forest);
+            const self = @fieldParentPtr(Replica, "state_machine", state_machine);
+            assert(self.committing);
+            assert(self.commit_callback != null);
+            assert(self.commit_prepare.?.header.op == self.op);
+            assert(self.commit_prepare.?.header.op == self.commit_min);
+            assert(self.commit_prepare.?.header.op == self.op_checkpoint_trigger());
+
+            // For the given WAL (journal_slot_count=8, lsm_batch_multiple=2, op=commit_min=7):
+            //
+            //   A  B  C  D  E
+            //   |01|23|45|67|
+            //
+            // The checkpoint is triggered at "E".
+            // At this point, ops 6 and 7 are in the in-memory immutable table.
+            // They will only be compacted to disk in the next measure.
+            // Therefore, only ops "A..D" are committed to disk.
+            // Thus, the SuperBlock's `commit_min` is set to 7-2=5.
+            const vsr_state_new = .{
+                .commit_min = self.op_checkpoint_next(),
+                .commit_max = self.commit_max,
+                .view_normal = self.view_normal,
+                .view = self.view,
+            };
+            assert(SuperBlock.VSRState.monotonic(self.superblock.working.vsr_state, vsr_state_new));
+
+            self.superblock.staging.vsr_state = vsr_state_new;
+            self.superblock.checkpoint(
+                commit_op_checkpoint_superblock_callback,
+                &self.superblock_context,
+            );
+        }
+
+        fn commit_op_checkpoint_superblock_callback(superblock_context: *SuperBlock.Context) void {
+            const self = @fieldParentPtr(Replica, "superblock_context", superblock_context);
+            assert(self.committing);
+            assert(self.commit_callback != null);
+            assert(self.commit_prepare.?.header.op == self.op);
+            assert(self.commit_prepare.?.header.op == self.commit_min);
+
+            self.op_checkpoint = self.op_checkpoint_next();
+            assert(self.op_checkpoint == self.commit_min - config.lsm_batch_multiple);
```

Comment:
I think we may trip this if checkpointing ever happens at `op_checkpoint_trigger = config.lsm_batch_multiple - 1`; this is the -1 edge case? We could do `-|` instead?

Reply:
I don't understand: checkpointing shouldn't trigger on `op_checkpoint_trigger == self.commit_min - config.lsm_batch_multiple` because checkpointing is triggered by comparing the `op_checkpoint_trigger == self.commit_min`.

Also,
- Right now `op_checkpoint_trigger()` is implemented as `self.op_checkpoint_next() + config.lsm_batch_multiple`.
- At this line, `self.op_checkpoint` was just advanced.
- `self.commit_min` is the "previous" `op_checkpoint_trigger()`. i.e. just before updating the `self.op_checkpoint` on line `2468`, `self.commit_min == self.op_checkpoint_trigger()`.

Reply:
At present, we would be fine.

I am thinking about if we optimize to checkpoint at the earliest possible checkpoint, for example, to accelerate startup after spin down, then for `batch_multiple=2`, we could have `op_checkpoint_trigger=1`, `commit_min=1`, `op_checkpoint_next=0` (which then stays the same i.e. 0 after the checkpoint at A):

```zig
   A  B
|01|23|45
```

And then if I understand correctly, `commit_min=1 - batch_multiple=2` would underflow the subtraction.

Reply:
We could also assert that a duplicate checkpoint like this never happens, for example, if `op_checkpoint=0` already then we shouldn't checkpoint at A because it wouldn't make a difference, since it's not helping with snapshotting the log—we're starting up reading from the same place in the log.

Reply:
> And then if I understand correctly, `commit_min=1 - batch_multiple=2` would underflow the subtraction.

I think the solution is to not trigger a checkpoint if nothing would actually be checkpointed.

> We could also assert that a duplicate checkpoint like this never happens, for example, if `op_checkpoint=0` already then we shouldn't checkpoint at A because it wouldn't make a difference, since it's not helping with snapshotting the log—we're starting up reading from the same place in the log.

So essentially `assert(op_checkpoint < op_checkpoint_next())`?

Reply:
Perfect.